Add tutorial: Compilation methods for Hamiltonian simulation circuits

henryzou50 opened the pull request:
# Add tutorial: Compilation methods for Hamiltonian simulation circuits

This PR adds a new tutorial to the Qiskit documentation titled **"Compilation methods for Hamiltonian simulation circuits"** and addresses issue #3552 

The notebook provides a comparative overview of three compilation methods in Qiskit for Hamiltonian simulation workloads:

- **SABRE transpiler**: Efficient routing and layout optimization.
- **AI Transpiler**: Intelligent, automated transpilation and synthesis for Pauli networks.
- **Rustiq plugin**: Precise and low-depth synthesis for `PauliEvolutionGate` operations used in Trotterized dynamics.

## What the tutorial covers
- Compilation strategies for circuits relevant to quantum simulation.
- Depth and gate count comparisons across methods.
- Circuit examples created from Qiskit’s Patterns module.

## Files added
- `docs/tutorials/100_representative.json`
- `docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb`

Note: The location of `100_representative.json` will likely need to be changed, as week as the notebook's reference of it. This json file is used as the file that contains the benchmarking circuits for Hamlib. What is the best location / most appropriate for this file?

## Before finalizing this tutorial
- Add a survey link to the end of the notebook (support appreciated).
- Update notebook metadata and JSON headers to follow Qiskit documentation conventions (support appreciated).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/documentation?pullRequest=3553) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/documentation?pullRequest=3553) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/documentation?pullRequest=3553) it.</sub>

christopherporter1 commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,2156 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/var/folders/jt/44gl1rhd26v6ykt0frqzl87r0000gn/T/ipykernel_59246/1325294152.py:17: MatplotlibDeprecationWarning: The get_cmap function was deprecated in Matplotlib 3.7 and will be removed in 3.11. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap()`` or ``pyplot.get_cmap()`` instead.\n",
```
Just flagging this warning, which I also get with all my installs updated. I'm not sure what the fixed syntax will be.

christopherporter1 commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,2156 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/var/folders/jt/44gl1rhd26v6ykt0frqzl87r0000gn/T/ipykernel_59246/1325294152.py:17: MatplotlibDeprecationWarning: The get_cmap function was deprecated in Matplotlib 3.7 and will be removed in 3.11. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap()`` or ``pyplot.get_cmap()`` instead.\n",
```
Just flagging this warning

henryzou50 replied:
Fixed as well!

christopherporter1 commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,2156 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/var/folders/jt/44gl1rhd26v6ykt0frqzl87r0000gn/T/ipykernel_59246/682414580.py:1: UserWarning: Providing `coupling_map` and/or `basis_gates` along with `backend` is not recommended, as this will invalidate the backend's gate durations and error rates.\n",
```
This warning appears several times. I don't understand why we need to pass basis gates to the pass manager. Won't it get that from the backend? I also don't know why we need this for the tutorial, because won't all the pms get the same set of basis gates from the backend? I do understand that the warning is not really important since this stops at transpilation and never runs anything. If you're sure that this step that throws the error is needed, ignore the comment.

henryzou50 replied:
Yes, usually we don't need to pass the basis gates to the pass manager, but I did that then as it was a temporary way to make the final gate types consistent with the results of the `ai_pauli` transpiler. Now we that updated, we can go back to how it usually is ran and we no longer have that warning.

christopherporter1 commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,2156 @@
+    }
+   ],
+   "source": [
+    "# QiskitRuntimeService.save_account(channel=\"ibm_quantum\", token=\"<MY_IBM_QUANTUM_TOKEN>\", overwrite=True, set_as_default=True)\n",
```
channel="ibm_quantum" is no longer valid on Cloud, right?

abbycross replied:
Correct! it should be ibm_quantum_platform for the channel

henryzou50 replied:
Correct and updated!

christopherporter1 commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,2156 @@
+    "- **How to leverage the AI transpiler for advanced, automated circuit optimization.**\n",
+    "- **How to employ the Rustiq plugin for circuits requiring precise synthesis of operations, particularly in Hamiltonian simulation tasks.**\n",
+    "\n",
+    "This tutorial uses three example circuits generated with Qiskit’s patterns module to illustrate the performance of each compilation method. By the end of this tutorial, users will be equipped to choose the appropriate compilation strategy based on their specific requirements and constraints.\n",
```
What is "Qiskit's patterns module"? I don't see the word "pattern" appear anywhere.

henryzou50 replied:
Oh I was referring to qiskit patterns. I updated the description and added a link to it.

christopherporter1 commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,2156 @@
+    "- The pie charts show that AI Pauli performed best for the largest percentage of circuits in terms of both depth and size.\n",
+    "- Despite this, Rustiq still had the lowest average depth and gate count.\n",
+    "\n",
+    "Each Hamiltonian circuit in this experiment differs significantly in structure, and this is reflected in the varying performance of the methods. The best-performing method depends on the circuit being compiled. Although Rustiq performed best on average, the results suggest that no single method dominates across all Hamiltonian circuits. We recommend users test multiple transpilation methods  ...
```
This is great. But I feel like I'm missing something. It seems the takeaway of all of this is "Use the AI transpiler" provided that you don't mind waiting for some potentially long transpilation times. But even that seems like it would almost always be the case: you would almost always trade classical overhead for quantum efficiency. Is there nuance that isn't being explicitly called out? Like are there cases where the scaling shown here breaks down, or could the message really be that broad: "Use AI transpiler"?

henryzou50 replied:
Thanks for this comment and yes, I agree, it does seem like that. I updated the analysis and further highlighted the limitations of the AI transpiler in that section.

christopherporter1 reviewed (commented):
Great tutorial. I learned a lot. Thank you. My comments are relatively minor.

henryzou50 commented:
Hi @abbycross and @christopherporter1 

Thanks so much for the helpful review! I’ve just pushed a commit that addresses the feedback. I also removed the `ai_pauli` transpiler for now — with the latest version of Qiskit and the transpiler service, it sometimes gets stuck in an infinite loop or errors out on certain Hamlib circuits. I’ll work with the AI transpiler team to figure out what’s going on there. In the meantime, the tutorial still works well for comparing the core methods: SABRE, AI, and Rustiq.

One thing I’m not too sure about now is how to handle the images since I’ve rerun all the experiments. Let me know if you have thoughts on the best way to update those.

abbycross commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,1652 @@
+    "### Compilation methods overview\n",
+    "\n",
+    "#### 1. **Qiskit transpiler with SABRE**\n",
+    "The Qiskit transpiler uses the SABRE (SWAP-based BidiREctional heuristic search) algorithm to optimize circuit layout and routing. SABRE focuses on minimizing SWAP gates and their impact on circuit depth while adhering to hardware connectivity constraints. This method is highly versatile and suitable for general-purpose circuit optimization, providing a balance between performance and comput ...
```
```suggestion
    "The Qiskit transpiler uses the SABRE (SWAP-based BidiREctional heuristic search) algorithm to optimize circuit layout and routing. SABRE focuses on minimizing SWAP gates and their impact on circuit depth while adhering to hardware connectivity constraints. This method is highly versatile and suitable for general-purpose circuit optimization, providing a balance between performance and computation time. To take advantage of the latest improvements in SABRE, detailed in [\\[1\\]](https://arxiv.org/abs/2409.08368), you can increase the number of trials (for example, `layout_trials=400, swap_trials=400`). For the purposes of this tutorial, we will use the default values for the number of trials in order to compare to Qiskit's default transpiler.  The advantages and parameter exploration of SABRE are covered in a separate [deep-dive tutorial](/docs/tutorials/transpilation-optimizations-with-sabre).\n",
```

abbycross commented on docs/tutorials/compilation-methods-for-hamiltonian-simulation-circuits.ipynb:
```diff
@@ -0,0 +1,1652 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAACw4AAAM9CAYAAACPb8ULAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQABAABJREFUeJzs3QmYVXX9P/DPLMCwI4vsgggookCKuKfkUpqaWZallLZYmmFl6s80tX4/NZWyzLSslCzzj5n7UppLuZRLiguQKAIyrIKAIOsw83/OoRlZZmAGLsPce16v57nPzL3nnDtn+XzPPct7vreoqqqqKgAAAAAAAAAAAACAgla8vWcAAAAAAAAAAAAAANj2BIcBAAAAAAAAAAA ...
```
```suggestion
    "- **AI transpiler** is typically reliable and provides depth-optimized circuits, but it comes with trade-offs in runtime, and also has other limitations, including supported coupling maps and synthesis capabilities. These are detailed in the [Qiskit Transpiler Service documentation](/docs/guides/qiskit-transpiler-service).\n",
    "  \n",
    "- In some cases, particularly with very large or hardware-specific circuits, the AI transpiler may not be as effective. In these cases, the default SABRE transpiler remains extremely reliable and can be further optimized by adjusting its parameters (see the [SABRE optimization tutorial](/docs/tutorials/transpilation-optimizations-with-sabre)).\n",
```

henryzou50 commented:
Thanks @abbycross for all the help! I just signed the CLA.
